condor_watch_q.py

```diff
@@ -420,6 +417,58 @@ def cli():
 GROUPBY_AD_KEY_TO_ATTRIBUTE = {v: k for k, v in GROUPBY_ATTRIBUTE_TO_AD_KEY.items()}
 
 
+def find_key_by_value(hash_map, value):
+    for k, v in hash_map.items():
+        if v == value:
+            return k
+    return None
+
+
+def group_clusters(
+    clusters,
+    key,
+    dagman_clusters_to_paths,
+    batch_names,
+    dagman_job_cluster_ids,
+    dag_nodes_total,
+):
+    groups = collections.defaultdict(list)
+    getter = operator.attrgetter(GROUPBY_AD_KEY_TO_ATTRIBUTE[key])
+    group_nodes_total = {}
+
+    for cluster in clusters:
+        # find cluster according to cluster path
+        dag_name = find_key_by_value(dagman_clusters_to_paths, cluster.event_log_path)
```

Comment:
I think the missing piece in simplifying this is this data structure:
```
dag_batch_names_by_nodes_log = {nodes_log_path: batch_names[cluster] for cluster, nodes_log_path in dagman_clusters_to_paths.items()}
```

You have two maps going in:
```
dagman_clusters_to_paths = {135: '/home/jovyan/1597258494/4/./test4.dag.nodes.log', 134: '/home/jovyan/1597258494/5/./test5.dag.nodes.log', 133: '/home/jovyan/1597258494/3/./test3.dag.nodes.log', 132: '/home/jovyan/1597258494/2/./test2.dag.nodes.log', 131: '/home/jovyan/1597258494/1/./test1.dag.nodes.log'}

batch_names = {105: 'test5.dag+105', 104: 'test4.dag+104', 103: 'test3.dag+103', 102: 'test2.dag+102', 101: 'test1.dag+101', 135: 'test4.dag+104', 134: 'test5.dag+105', 133: 'test3.dag+103', 132: 'test2.dag+102', 131: 'test1.dag+101'}
```

You want a map that goes from the values of first map to the values of the second. The problem is that the values of both maps are not unique, so you can't just invert the first map and then do a lookup through both mappings.

Your current implementation looks for any key in the first map that has the right value, then passes that to the second map. This is fine, because the second map also has lots of duplicate information in it. However, it is very inefficient because it has to do a linear scan on every lookup, and produces odd code besides.

You can bypass the scan by building the above data structure, which ends up looking like this:
```
{'/home/jovyan/1597258494/4/./test4.dag.nodes.log': 'test4.dag+104', '/home/jovyan/1597258494/5/./test5.dag.nodes.log': 'test5.dag+105', '/home/jovyan/1597258494/3/./test3.dag.nodes.log': 'test3.dag+103', '/home/jovyan/1597258494/2/./test2.dag.nodes.log': 'test2.dag+102', '/home/jovyan/1597258494/1/./test1.dag.nodes.log': 'test1.dag+101'}
```

The key realization is that the nodes-log-to-batch-name mapping **is** 1:1, but you only have it in terms of this messy middle layer of the two maps. The dictionary comprehension ends up setting the same key-value pair multiple times, but it doesn't matter because it sets it to the same thing each time, using the cluster ID as the (multi-valued) bridge between the two. Point being: this functionally does the same thing as what your code does, but it pre-builds the mapping so that it doesn't have to do a linear scan each time, and results in (I think) a clearer function.

The revised function looks like this:
```
def group_clusters(
    clusters,
    key,
    dagman_clusters_to_paths,
    batch_names,
    dagman_job_cluster_ids,
    dag_nodes_total,
):
    groups = collections.defaultdict(list)
    getter = operator.attrgetter(GROUPBY_AD_KEY_TO_ATTRIBUTE[key])
    group_nodes_total = {}

    dag_batch_names_by_nodes_log = {nodes_log_path: batch_names[cluster] for cluster, nodes_log_path in dagman_clusters_to_paths.items()}

    for cluster in clusters:
        # find cluster according to cluster path
        dag_batch_name = dag_batch_names_by_nodes_log.get(cluster.event_log_path)

        if key == BATCH_NAME and dag_batch_name is not None:
            groups[dag_batch_name].append(cluster)
            group_nodes_total[dag_batch_name] = dag_nodes_total.get(
                dag_batch_name
            )
        else:
            if cluster.cluster_id not in dagman_job_cluster_ids:
                groups[getter(cluster)].append(cluster)

    if not dag_nodes_total:
        group_nodes_total = dag_nodes_total

    return groups, group_nodes_total
```

The `find_key_by_value` helper is then unnecessary and can be deleted.